### ISSUE TITLE
Not working correctly with Gollum 3.1

### ISSUE THREAD

**Issue by neeravkumar** [arr2036/omnigollum#24]
Hi,

With Gollum 3.1, I can login but don't have any edit privileges.

Have you seen this too?

**Comment by dometto** [comment]
The idea was indeed to make `--no-edit` optional, but a bug sneaked in: for those not using gollum via the `gollum` command on the CLI, but by mounting `Precious::App` in some other way, the `@allow_editing` option would accidentally default to true. See [here](https://github.com/gollum/gollum/issues/911). We fixed this in gollum `v3.1.1`, so `gem update gollum` should fix the problem. To be clear: this is not an omnigollum bug. Sorry for the confusion!

**Comment by explody** [comment]
Following on the reference to this issue that I made in a gollum issue report (noted above), we thought we'd bounce it over here.  

For me, it appears that in ```editing_auth.rb```, ```@app``` is somehow arriving as a OmniAuth::Builder object and ```@app.settings``` is private.

It's unclear to me where it core issue lies or whether it's directly related to this ```@allow_editing``` issue.  If you'd like a separate ticket for it, I'll be happy to do so.

**Comment by dometto** [comment]
Sorry to add to the confusion, but it turns out that https://github.com/gollum/gollum/issues/911 was not resolved completely (read all about it [here](https://github.com/gollum/gollum/pull/950)). I just merged and released the proper fix in gollum `v3.1.2`.

@explody, I think this is probably unrelated to your issue, but you could give it a try again with the patch just released.

**Comment by techwiz24** [comment]
@dometto Sinatra is bombing out (`undefined method 'join' for #<String:0x...>`) with `v3.1.2` when trying to login via LDAP w/ `omnigollum`, so I don't think it's fixed. [A quick search](https://stackoverflow.com/questions/26255022/why-is-sinatras-show-exceptions-rb-file-crashing-when-i-raise-a-runtimeerror) says that it may be a `rack` bug conflicting with `sinatra`, which is masking the actual exception. According to sinatra/sinatra#907,an update to sinatra should be coming soon to correct this, but for now, I'm stuck on `3.1.1`.

**Comment by dometto** [comment]
@techwiz24, annoying that it's not showing the real cause, no way to get a more informative backtrace? I'm getting the `join` issue when I e.g. try to point gollum to a non-existent repo, so it might be a small issue like that.

However, I'm not getting the error when running `rackup config.ru` with sane values for the wiki options. So it wouldn't surprise me if the underlying issue for you is the same as for @explody (see https://github.com/gollum/gollum/issues/940). If you want to try and debug you could try changing this line:
https://github.com/gollum/gollum/blob/master/lib/gollum/editing_auth.rb#L10
To something like:
```ruby
unless true
```
So that the call to `.settings` is removed.

**Comment by aldanor** [comment]
@techwiz24 you'll see exceptions if you use rack v1.5.2

**Comment by techwiz24** [comment]
@aldanor Thanks, I'll have to check that the next time I'm in the office. However, it still seems sinatra/sinatra#907 hasn't closed been closed yet. As I said before, I'm not really sure where the base issue lies (Gollum/Omnigollum/Rack/Sinatra/something else), but hopefully we'll be able to hunt this down!

**Comment by nebulade** [comment]
I am getting the same error like @explody 
```
private method `settings' called for #<OmniAuth::Builder:0x00000003b41ad8>
```
I am using gollum (3.1.2) and omnigollum (0.1.4)

**Comment by techwiz24** [comment]
Hmm, Same problem as @nebulade and @explody. Can't seem to log in anymore via omnigollum.

These are the gem versions I have:

```
gollum (3.1.2)
gollum-grit_adapter (0.1.3)
gollum-lib (4.0.2)
sinatra (1.4.6)
rack (1.6.0)
rack-contrib (1.2.0)
rack-protection (1.5.3)
```

Now that I'm back in the office, I can finally get some time to fiddle with this (hopefully!).

**Comment by dometto** [comment]
The core issue is the `private method 'settings'` one. The `join` issue disappears when upgrading to sinatra `v1.4.6`.

The core problem is caused by the presence of the new `EditingAuth` middleware in gollum (see [here](https://github.com/gollum/gollum/issues/940)). That middleware is expecting the app it is passed to be Gollum, but when using omnigollum, it's in fact handed an instance of `OmniAuth::Builder`. That's done by omnigollum [here](https://github.com/arr2036/omnigollum/blob/master/lib/omnigollum.rb#L217). Not sure how we can solve this: might the problem be that the two middlewares (`OmniAuth::Builder` and `Precious::EditingAuth`) are currently being added in the wrong order? If so, is there any way to enforce the correct order?

**Comment by dometto** [comment]
It does work fine when I add the `EditingAuth` middleware *after* omnigollum puts `OmniAuth::Builder` on the middleware stack. However, to do so, I need to do `Precious::App.use Precious::EditingAuth` outside of `Precious:App`. That won't work when using gollum as a rack app (e.g. with `config.ru` and `rackup`), because then `EditingAuth` won't get used at all unless the user manually adds it.

Bottom line: is there a way for omnigollum to ensure that `OmniAuth::Builder` is put on the middleware stack in front of the other middleware? I don't see a way to change the ordering of middleware in Sinatra. Sinatra's `use` just appends the middleware to the stack.

**Comment by dometto** [comment]
Re-implemented the blocking of non-GET requests without the piece of middleware in Re-implemented the blocking of non-GET requests without the piece of middleware. That resolves the problem for me locally. That should resolve this (and related) issues.

**Comment by techwiz24** [comment]
@dometto awesome! Any idea when the next release will be cut? I'd love to get the latest changes into the wiki we use at work (We need Omnigollum for LDAP).

**Comment by dometto** [comment]
@techwiz24 can you confirm that everything is ok on gollum `HEAD`? Then I'll do a bugfix release tonight.

**Comment by techwiz24** [comment]
@dometto Seems to be fixed in `HEAD` (*6d33687*), as I can successfully log in now using OmniGollum and LDAP! Thanks!

**Comment by dometto** [comment]
I've released 3.1.3, containing the fix. :) @arr2036, you can close this and the other two related issues. Sorry for the trouble.

**Comment by arr2036** [comment]
@dometto thanks :)
